### ISSUE TITLE
Starting in 1.25 clusters, services of type=LB and xTP=Local sometimes does not update node backends on load balancers 

### ISSUE THREAD

**Issue by swetharepakula** [kubernetes/kubernetes#112793]
### What happened?

When upgrading nodes from 1.24 to 1.25, on a cluster where master is already at 1.25, I notice that my Service `type=LoadBalancer` and `xTP=Local` have an incorrect set of nodes after the nodes have been upgraded. The set contains only the old nodes that no longer exist resulting my service being unavailable through my load balancer.

### What did you expect to happen?

I would expect that the load balancer to be properly updated with the new set of nodes after the upgrade.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create 1.25 Cluster with 1.24 nodes
2. Deploy a service of `type=LoadBalancer` and `xTP=Local`
3. Upgrade the 1.24 nodes to 1.25
4. After upgrade is finished look at the node list for the Load Balancer.

### Anything else we need to know?

The existing logging is not enough to diagnose the issue. I added some more logs and ran the KCM at log level =5 to find the root cause.

There was a change introduced to reduce the number syncs for `xTP=Local` services: #109706. With this change, there are situations where the `xTP=Local` never gets updated. 

The following is the chain of events.
1. Node is created or deleted and causes `triggerNodeSync()`:

https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L169-L192

2. Inside `nodeTriggerSync()`, the nodeLister (line 264) filters for Ready only nodes so it does not have the new node or still contains the deleted node, which means that `c.needFullSync = false` when line 281 is executed.

https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L260-L288

3.  Following down the chain of functions that are called (across goroutines communicating with nodeSyncCh), we end up at `nodeSyncInternal`. Because `c.needFullSync = false`, we will only do a  sync of services that were marked for retry. If the state previously was good, this means `c.servicesToUpdate` has 0 services before entering `updateLoadBalancerHosts`.
https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L725-L741

4.  Nodes are queried again from the NodeLister. But this time the new node or the deleted node is reflected. 
https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L782-L811

5. nodeSyncService is then parallelized based on the length of services. In this case we have no services, so we do no updates but on line 808, c.lastSyncedNodes is set to the nodes found in step 4.
https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L808

A subsequent full sync does not fix things. We go through steps 1 through 5 again, the difference being `c.needFullSync=true`. This will mean that in step 4 & 5, `c.servicesToUpdate` will not be empty resulting in the following:

6. inside `nodeSyncService` we filter based on predicates for the `xTP=Local` and `xTP=Cluster`. In the case of `xTP=Local` since we do not pay attention to `Ready` status, all of the nodes in the `c.lastSyncedNodes` will be the `oldNodes`. And since no node creations or deletions have occurred, all the `newNodes` will be the same. This results in no sync (line 767).

https://github.com/kubernetes/kubernetes/blob/a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L759-L777

In the the `xTP=Cluster`, the node ready status is used to filter the nodes, so `c.LastSyncedNodes` would already have the newly created node, but it would not be ready. Which means the node will not be part of `oldNodes` but it will exist in `newNodes` and allows the sync to continue as expected.
 

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2", GitCommit:"5835544ca568b757a8ecae5c153f317e5736700e", GitTreeState:"clean", BuildDate:"2022-09-21T14:33:49Z", GoVersion:"go1.19.1", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.2-gke.300", GitCommit:"6f9a8e57036ff71785ef9c90998437413a3a8ff5", GitTreeState:"clean", BuildDate:"2022-09-26T09:26:16Z", GoVersion:"go1.19.1 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Tested on GKE, however this should theoretically be possible on any 1.25+ cluster.
</details>

### OS version

N/A

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by swetharepakula** [comment]
I have done a test with code on master and looked through the changes that have been made since 1.25 was cut and this issue seems to be fixed.  I believe this issue is specific only to 1.25

**Comment by alexanderConstantinescu** [comment]
/assign 

I filed https://github.com/kubernetes/kubernetes/pull/112798 which provides an in-depth answer as to why that fixes the problem

**Comment by liggitt** [comment]
with https://github.com/kubernetes/kubernetes/pull/112807 merged, is this resolved?

**Comment by swetharepakula** [comment]
This is resolved with  #112807

/close

**Comment by JoelSpeed** [comment]
@alexanderConstantinescu The revert that you put up in #112807 went directly into the release-1.25 branch. I believe I am seeing the same issue again on release-1.26 and release-1.27. Did a revert or fix go up to the master branch that I've missed?

**Comment by alexanderConstantinescu** [comment]
@JoelSpeed: There was no revert PR needed on master, because master (at that time) didn't have the bug. This is the PR I filed which attempted to fix the issue on 1.25: https://github.com/kubernetes/kubernetes/pull/112798 but which later got superseded by the revert PR. It explains the issue in greater depth. 

What issue are you seeing?

**Comment by JoelSpeed** [comment]
Observing on Azure, with the in-tree cloud provider in 1.27 (this could be a bug directly in the Azure code to be fair but wanted to rule this bug out too)
* KCM boots up on new cluster
* New worker nodes are added to the cluster, CSRs approved etc
* Workers do not get added to the LB on first pass because they are not ready (Azure specific? We are using `stableNodeSetPredicates`)
* Workers eventually go ready, but `EnsureLoadBalancer` never called again

The symptoms look the same as this bug, but I'm guessing now that we are using `stableNodeSetPredicates`, the Azure provider cannot exclude nodes based on their readiness, since the service controller will no longer tell it to update the nodes in the load balancer for an `eTP: Local` service?

**Comment by JoelSpeed** [comment]
I think I found the issue, https://github.com/kubernetes-sigs/cloud-provider-azure/issues/4230

**Comment by liggitt** [comment]
> @JoelSpeed: There was no revert PR needed on master, because master (at that time) didn't have the bug. This is the PR I filed which attempted to fix the issue on 1.25: #112798 but which later got superseded by the revert PR. It explains the issue in greater depth.

The revert PR only landed in the release-1.25 branch, and #112798 never merged... what resolved this issue in master and 1.26+?

**Comment by alexanderConstantinescu** [comment]
> what resolved this issue in master and 1.26+?

This bug never existed on master / 1.26 to begin with. Here's a timeline of events:

1. [1.25 feature cycle]: PR https://github.com/kubernetes/kubernetes/pull/109706 merges on 1.25 with a bug 
2. [1.26 feature cycle]: PR https://github.com/kubernetes/kubernetes/pull/111663 merges on 1.26 with the fix to said bug (the bug was not known at the time, and this PR was originally created as a cleanup but ended up also fixing the bug)
3. [1.25 code freeze cycle]: PR https://github.com/kubernetes/kubernetes/pull/112807 merges on 1.25, reverting 1.

I.e: we could have applied https://github.com/kubernetes/kubernetes/pull/112798 in step 3, but chose to revert instead

Let me know if this is not clear?
